### ISSUE TITLE
Error transpiling private static final fields in private classes

### ISSUE THREAD

**Issue by andfaulkner** [cincheo/jsweet#240]
Any private static final field in a private class gets transpiled such that in the resultant output, the value of the field (as declared in Java) gets assigned to the instance rather than to the class. When compiling from Typescript to Javascript, this causes "Property not found" errors and can crash the Typescript build. Also, it makes these values unavailable on the class object.

# Example

Java input:

    public class Transaction implements Serializable {
        // .. //

        /**
         * Different action types this transaction can be
         */
        private class ActionType {
            private static final int CREATE = 0;
            private static final int UPDATE = 1;
            private static final int DELETE = 2;
        }
    }

Typescript output:

    export namespace Transaction {
        /**
         * Different action types this transaction can be
         */
        export class ActionType {
            public __parent: any;
            static CREATE : number;

            static UPDATE : number;

            static DELETE : number;

            constructor(__parent: any) {
                this.__parent = __parent;
                this.CREATE = 0;
                this.UPDATE = 1;
                this.DELETE = 2;
            }
        }
        ActionType["__class"] = "ca.ohri.javelin.sync.model.Transaction.ActionType";
    }

Later attempting to transpile this code into Javascript results in the following errors:

    181             this.CREATE = 0;
                         ~~~~~~
    [*]/Transaction.ts(181,18): error TS2339: Property 'CREATE' does not exist on type 'ActionType'.

    182             this.UPDATE = 1;
                         ~~~~~~
    [*]/Transaction.ts(182,18): error TS2339: Property 'UPDATE' does not exist on type 'ActionType'.

    183             this.DELETE = 2;
                         ~~~~~~
    [*]/Transaction.ts(183,18): error TS2339: Property 'DELETE' does not exist on type 'ActionType'.

# Proposed fix
This would be remedied if JSweet instead assigned these static properties to the class rather than the instance. For example:

    export namespace Transaction {
        /**
         * Different action types this transaction can be
         */
        export class ActionType {
            public __parent: any;
            static CREATE : number;

            static UPDATE : number;

            static DELETE : number;

            constructor(__parent: any) {
                this.__parent = __parent;
                ActionType.CREATE = 0;   // <<<<< fixes issue
                ActionType.UPDATE = 1;   // <<<<< fixes issue
                ActionType.DELETE = 2;   // <<<<< fixes issue
            }
        }
        ActionType["__class"] = "ca.ohri.javelin.sync.model.Transaction.ActionType";
    }

**Comment by renaudpawlak** [comment]
Ouch. That's a bad one :( I can't believe I did not hit it before. Thanks a lot for reporting.

Although it probably fix the issue for private classes, I am reluctant with initializing the static fields in the constructor because it is not the right place to do it (those fields should be accessible without having to instantiate the class). I'd prefer dealing this these fields like any other static field (actually they should but there is a weird bug that might be tricky here).

I'll look at is when possible.

**Comment by andfaulkner** [comment]
Thanks @renaudpawlak!

You're right about initialization not belonging in the constructor. The better solution would be to use static property initializers:

        /**
         * Different action types this transaction can be
         */
        export class ActionType {
            public __parent: any;

            public static CREATE : number = 0;
            public static UPDATE : number = 1;
            public static DELETE : number = 2;

            constructor(__parent: any) {
                this.__parent = __parent;
            }
        }
        ActionType["__class"] = "ca.ohri.javelin.sync.model.Transaction.ActionType";
    }

...which it sounds to me like you're suggesting anyway.

**Comment by renaudpawlak** [comment]
Yes. The point is that since it already works for regular classes, it might be an issue with inner classes (by the way have you tried with non-private inner classes?). IMO the fix should just make sure that the static fields behave the normal way, even for inner private classes. I'll check.

**Comment by renaudpawlak** [comment]
Thanks! It was quite easy to fix once this bug was well identified. Please re-open if not fixed in 2.0.0-SNAPSHOT.

**Comment by andfaulkner** [comment]
Fantastic, thanks very much @renaudpawlak
